### ISSUE TITLE
Changelog?

### ISSUE THREAD

**Issue by dwhswenson** [openpathsampling/openpathsampling#641]
One of the recommendations in the E-CAM project is that we use a proper change log. I'm unsure whether we need this, since I think that the git repository and GitHub serve this role for us. When I need the change log, I either use git commands or I browse the merged PRs on GitHub (for example, when doing release notes).

I think the PRs give a better picture of what has changed than the change log would, and I think a change log would be a lot of maintenance for rapidly changing code (the change log would soon be longer than the code itself.) However, I'm not totally opposed to the idea. One possibility might be to maintain a change log starting with 1.0 and through our 1.x series?

Putting this issue up for discussion; it was listed as something to be included for the E-CAM modules, but our detailed PRs might meet the same needs. (I think we take the attitude that a public GitHub acts as additional documentation, as well as a collaborative development platform.)

Thoughts, @jhprinz?

**Comment by jhprinz** [comment]
Well, the changeling are a one file summary of what has changed in PRs? Yes? Or even the changes per release. It makes sense to keep this. Most people will not browse closed PRs and see what changed. 

We could keep another changeling for the current master to let developers know when changes happened. Once we cut a release we just copy the info? Would that make sense?

**Comment by dwhswenson** [comment]
My understanding of what the E-CAM team suggests is per-PR updates, following the format at https://en.wikipedia.org/wiki/Changelog#Format. We should consider the discussion at http://keepachangelog.com/ as well.

An additional `NEWS` file to help write release notes could also be useful (as a summary of the change log summary). That would be per-release.

I think if we do anything like this, it should start after 1.0. I feel like starting before 1.0 implies that we should do it retroactively, which is just way too much work.

**Comment by jchodera** [comment]
Can you point us to the E-CAM written guidelines?

**Comment by jchodera** [comment]
Relatedly, we are trying to collect our standard best practices (intended for new students) here:
https://github.com/choderalab/software-development

Comments, feedback, and contributions are welcomed!

**Comment by jhprinz** [comment]
> I think if we do anything like this, it should start after 1.0. I feel like starting before 1.0 implies that we should do it retroactively, which is just way too much work.

Definitely after 1.0. But we have now several people developing so it would be nice to have some way for them to track the PRs we do. We could start at 0.9.1

**Comment by jchodera** [comment]
A changelog is usually just a summary of important changes and bugfixes between versions for users. I don't think anything before 1.0 is relevant, since 1.0 is the first real public release to "users" (and not just people who live on the bleeding edge and are OK with things being changed around without warning).

**Comment by jchodera** [comment]
MDTraj has a good example: http://mdtraj.org/1.8.0/whatsnew.html

**Comment by dwhswenson** [comment]
> A changelog is usually just a summary of important changes and bugfixes between versions for users.
> MDTraj has a good example: http://mdtraj.org/1.8.0/whatsnew.html

I think there are two different kinds of documents under discussion here. This is one of the interpretations, and I fully agree that we should have good "what's new" release notes like this. However, this doesn't seem to be the definition that E-CAM is using (although we might get some feedback from the E-CAM SW team to see exactly what they intend -- I've emailed the lead on the SW team with a link to this issue, and I hope he'll provide insight.)

What I read in the E-CAM documentation are the following two points:

* each E-CAM "module" should involve an update to the change log
* the change log should be in the GNU changelog format at https://en.wikipedia.org/wiki/Changelog#Format

That format indicates to me that each PR is a heading, and there are several bullet points per PR, in contrast to one-bullet-per-PR with headings per-release in the "what's new" file.

For context, I think an E-CAM "module" will typically consist of a few PRs -- my proposed modules seem to be between 1-5 PRs. (You can see the proposed Classical MD modules at https://gitlab.e-cam2020.eu/E-CAM/Classical-MD-Modules/merge_requests/. "Modules" in that repo are files in rst format describing the code that has been added. As of this writing, none have been merged and approved; the in-progress modules include 6 that I've written up and 6 from our ESDW participants.)

Anyway, my concern is that a bullet point per function changed (as I understand the GNU format) is overwhelming (especially on a fast-moving project like this), a burden to maintain, and should be redundant with well-written PR descriptions. On the other hand, one bullet point per PR should not cause any problems.

**Comment by jchodera** [comment]
Seriously? The linked GNU style [is found here](https://www.gnu.org/prep/standards/html_node/Style-of-Change-Logs.html).

That's the peak of 1998 LISP-coding style.

Good luck, and may god have mercy on your souls.

**Comment by dwhswenson** [comment]
Well, this is why I'm asking Alan to clarify. This has happened a few times before (over the coding guidelines conflicting with PEP8, and over Doxygen being explicitly required, which conflicts with our Sphinx/numpydoc style). What I see is that the SW team cares about the big picture of writing good software, and trying to train scientific software developers to use best practices that scientists often ignore. I expect that their advice here will also be to do something that meets the essential goal of summarizing changes for more casual contributors, while not drowning us in extra metadata overhead.

But before we start managing a changelog, I want to make sure it will meet the E-CAM guidelines.

**Comment by ocaisa** [comment]
Hi all, I'm Alan from the E-CAM here. I think David has hit the nail on the head in his last comment. We just want people to be aware of best practices, why they are best practices and to start employing (some) of them if they aren't already. 
Our guidelines are very much a work in progress that have been cobbled together from a number of other sources. The plan is to make them available online and to tune them up there. At this stage we have very few obligatory items: use of a style guide, source code documentation and testing. Something like this thread is something that would be great to see as a discussion when our documentation makes it online, I would be more than happy to lighten the language and provide better context.

**Comment by dwhswenson** [comment]
Thanks @ocaisa!

So it sounds like the best thing to do would be do a heading-per-release, bullet-point-per-PR type of changelog, similar to MDTraj's "what's new" file. I'd rather start that as of 1.0, saying "what's new in 1.0? Everything!" (Although any other significant changes after 0.9.1 should be highlighted in release notes in the 0.9.x series.)

If someone wants to search the history the way the GNU changelog format enables, they can do that using tools built into git. A little GitHub magic (pasting the commit hash into the search field) and they can find the relevant PR, including all the detailed conversation about why we did what we did.

One practice that I hadn't considered before is that, starting with 1.0, each PR that needs to add something to the what's new/release notes should include that in the changelog. This seems to be part of what the E-CAM guidelines suggest, and it would make writing release notes a *lot* easier.

**Comment by jchodera** [comment]
@ocaisa : It will be great to see the draft standards online when they're posted! Our lab (which is very Python-centric) has started to collect some of our best practices in draft form online, with the hope of roping in more groups and evolving toward a standard set of guidelines that make it easier to collaborate and contribute to each others codebases:
https://github.com/choderalab/software-development

**Comment by ocaisa** [comment]
I'll get on it early in the new year...but I took at yours and you are already more developed than what we have! Most of the targets for me at the moment are in the Fortran space, there's not a lot of (easy to find) good advice out there (unlike Python). I also need to encourage people to think HPC from the outset.

Back on the subject of changelog, there's a generator available from PRs which would do a pretty good job (and the approach is something I've seen on other projects I'm involved in): https://github.com/skywinder/Github-Changelog-Generator
Of course it assumes a certain workflow....

**Comment by dwhswenson** [comment]
Closing this issue. Our change log is the release notes we write with each release. The first draft of those release notes are automatically generated by a tool I've written that gathers all the PRs since the last release, sorts them according to their tags to determine whether they're "New features", "Bugs fixed", or "Miscellaneous improvements", and then writes a draft of the main release notes, including tagging authors of PRs (other than me and @jhprinz). I've been using that tool since circa 0.9.4, I think (and using it on other projects, too), so I feel like it's pretty trustworthy now.

**Comment by jchodera** [comment]
> The first draft of those release notes are automatically generated by a tool I've written that gathers all the PRs since the last release, sorts them according to their tags to determine whether they're "New features", "Bugs fixed", or "Miscellaneous improvements", and then writes a draft of the main release notes, including tagging authors of PRs (other than me and @jhprinz). 

Sounds like a useful tool! Where does this live?

**Comment by dwhswenson** [comment]
> Sounds like a useful tool! Where does this live?

It's in a PR for my autorelease package: https://github.com/dwhswenson/autorelease/pull/29. It hasn't been merged into master yet because it was doing enough for my needs, and until today I was too lazy to go back and figure out why AppVeyor wasn't passing. I'll probably merge and cut a new release of autorelease soon.

The release note writer uses PR labels for the headings of the release notes (e.g., "New features" or "Bugs fixed"). You need to manually add those tags to PRs. The way I use it, the code writes the first draft of the release notes. Then I edit those to be clearer. Sometimes PR titles aren't great; sometimes one idea was implemented through several PRs; and sometimes one PR has multiple purposes. Also, it will remind you of any PRs that have been merged without any labels, so you can either label the PR and re-run, or just edit the release notes.

That whole package is a way of managing releases, a bit inspired by gitflow, but with some differences in the particulars. I've been using it for my [contact map](https://github.com/dwhswenson/contact_map) package, and, of course, for autorelease itself. Once I feel more confident about the autorelease tools, I'd like to use them for the OPS `setup.py`, `travis.yml`, and release cycle.

The `write-release-notes` script takes a yaml configuration file as `--conf file.yaml`. This sets things like default contributors and headings based on PR labels. Here's the current yaml for OPS (with my GitHub token removed; I haven't committed this yaml file to the repo because I haven't separated authentication from customization yet):

```yaml
# authentication
github_user:
    username: dwhswenson
    token: REMOVED

# repo identification
project:
    repo_owner: openpathsampling
    repo_name: openpathsampling
    project_name: OpenPathSampling

# PR labels/headers for output
labels:
    - label: feature
      heading: New features
    - label: bugfix
      heading: Bugs fixed
    - label: misc PR
      heading: Miscellaneous improvements
      topics:
          - label: docs
            name: Improvements to documentation
          - label: code-style
            name: Improvements to code readability and style
          - label: internal
            name: Behind-the-scenes improvements to CI, testing, and deployment

# standard contributors don't get mentioned in release notes
standard_contributors:
    - dwhswenson
    - jhprinz
```
